backblazeb2/src/main/scala/akka/stream/alpakka/backblazeb2/scaladsl/B2API.scala

```diff
@@ -0,0 +1,235 @@
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.Uri.Query
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.unmarshalling.{FromResponseUnmarshaller, Unmarshal}
+import akka.stream.Materializer
+import akka.stream.alpakka.backblazeb2.B2Encoder
+import akka.stream.alpakka.backblazeb2.Protocol._
+import akka.stream.alpakka.backblazeb2.SerializationSupport._
+import akka.util.ByteString
+import cats.data.EitherT
+import cats.syntax.either._
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.concurrent.Future
+
+object B2API {
+  val DefaultHostAndPort = "api.backblazeb2.com"
+}
+
+/**
+ * Handles the interface with the Backblaze B2 API, but doesn't handle any expired tokens or retries
+ */
+class B2API(hostAndPort: String = B2API.DefaultHostAndPort)(implicit system: ActorSystem, materializer: Materializer)
+    extends LazyLogging {
+  implicit val executionContext = materializer.executionContext
+  private val version = "b2api/v1"
+
+  /**
+   * https://www.backblaze.com/b2/docs/b2_authorize_account.html
+   */
+  def authorizeAccount(credentials: B2AccountCredentials): B2Response[AuthorizeAccountResponse] = {
+    val encodedCredentials = encodeBase64(s"${credentials.accountId}:${credentials.applicationKey}")
+    val authorization = s"Basic $encodedCredentials"
+    val request = HttpRequest(
+      uri = s"https://$hostAndPort/$version/b2_authorize_account",
+      method = HttpMethods.GET
+    ).withHeaders(RawHeader("Authorization", authorization))
+
+    requestAndParse[AuthorizeAccountResponse](request)
+  }
+
+  /**
+   * https://www.backblaze.com/b2/docs/b2_get_upload_url.html
+   */
+  def getUploadUrl(
+      authorizeAccountResponse: AuthorizeAccountResponse,
+      bucketId: BucketId
+  ): B2Response[GetUploadUrlResponse] = {
+    val apiUrl = authorizeAccountResponse.apiUrl
+    val accountAuthorization = authorizeAccountResponse.authorizationToken
+    val uri = Uri(s"$apiUrl/$version/b2_get_upload_url").withQuery(Query("bucketId" -> bucketId.value))
+    val headers = RawHeader("Authorization", accountAuthorization.value)
+    val request = HttpRequest(
+      uri = uri,
+      method = HttpMethods.GET
+    ).withHeaders(headers)
+
+    requestAndParse[GetUploadUrlResponse](request)
+  }
+
+  /**
+   * https://www.backblaze.com/b2/docs/b2_upload_file.html
+   */
+  def uploadFile(
+      uploadCredentials: GetUploadUrlResponse,
+      fileName: FileName,
+      data: ByteString,
+      contentType: ContentType = DefaultContentType
+  ): B2Response[UploadFileResponse] = {
+    val uri = Uri(uploadCredentials.uploadUrl.value)
+    val headers =
+    RawHeader("Authorization", uploadCredentials.authorizationToken.value) ::
+    RawHeader("X-Bz-File-Name", B2Encoder.encode(fileName.value)) ::
+    RawHeader("X-Bz-Content-Sha1", B2Encoder.sha1String(data)) ::
+    Nil
+
+    val entity = HttpEntity(data).withContentType(contentType)
```

Comment:
I would prefer here `data` to be a `Source[ByteString, ...]`. This would allow a stream of `ByteString`s to be sent to Backblaze B2 using chunked encoding.

It is always better to natively support the stream use-case as it is always possible to use an implementation that uses streams to upload strict (not-streamed) data. But other way around - not so.